File: mobile/lib/core/services/secure_storage_service.dart
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import '../constants/app_constants.dart';

class SecureStorageService {
  SecureStorageService({FlutterSecureStorage? storage})
      : _storage = storage ?? const FlutterSecureStorage();

  final FlutterSecureStorage _storage;

  Future<void> saveToken(String token) =>
      _storage.write(key: AppConstants.secureStorageTokenKey, value: token);

  Future<String?> readToken() => _storage.read(key: AppConstants.secureStorageTokenKey);

  Future<void> clearToken() => _storage.delete(key: AppConstants.secureStorageTokenKey);

  Future<bool> hasCompletedOnboarding() async {
    final value = await _storage.read(key: AppConstants.secureStorageOnboardingKey);
    return value == 'true';
  }

  Future<void> markOnboardingComplete() =>
      _storage.write(key: AppConstants.secureStorageOnboardingKey, value: 'true');
}

File: appReact/src/core/services/secureStorageService.ts
import * as SecureStore from 'expo-secure-store';
import { SECURE_STORAGE_TOKEN_KEY, SECURE_STORAGE_ONBOARDING_KEY } from '../constants/appConstants';

export interface SecureStoreAdapter {
  setItemAsync(key: string, value: string): Promise<void>;
  getItemAsync(key: string): Promise<string | null>;
  deleteItemAsync(key: string): Promise<void>;
}

export class SecureStorageService {
  private storage: SecureStoreAdapter;

  constructor(storage: SecureStoreAdapter = SecureStore) {
    this.storage = storage;
  }

  saveToken(token: string): Promise<void> {
    return this.storage.setItemAsync(SECURE_STORAGE_TOKEN_KEY, token);
  }

  readToken(): Promise<string | null> {
    return this.storage.getItemAsync(SECURE_STORAGE_TOKEN_KEY);
  }

  clearToken(): Promise<void> {
    return this.storage.deleteItemAsync(SECURE_STORAGE_TOKEN_KEY);
  }

  async hasCompletedOnboarding(): Promise<boolean> {
    const value = await this.storage.getItemAsync(SECURE_STORAGE_ONBOARDING_KEY);
    return value === 'true';
  }

  markOnboardingComplete(): Promise<void> {
    return this.storage.setItemAsync(SECURE_STORAGE_ONBOARDING_KEY, 'true');
  }
}
